health: Split out passive mode into separate binary

gandro opened the pull request:
This change introduces a new binary called `cilium-health-probe`, which replaces the 'passive' mode the existing `cilium-heath` executable. Its purpose is emulating an endpoint by running a HTTP server in the cilium-health namespace for testing endpoint connectivity.

To reduce memory consumption, the new TCP-only server in the new binary is based on net/http instead of go-openapi and comes with a small client. The old `/hello` endpoint of the cilium-health API is still there, as it is also accessed by commands such as `cilium-health ping` on the "admin API" (exposed via unix socket only), but I plan to eventually remove it in a follow-up PR.

The new binary reduces the resident set size (RSS) of the endpoint process by roughly 12-15MB.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8317)
<!-- Reviewable:end -->

gandro committed:
health: Split out passive endpoint into separate binary

This change introduces a new binary called 'cilium-health-responder',
which replaces the 'passive' mode the existing cilium-heath executable.
Its purpose is emulating an endpoint running a HTTP server in the
cilium-health namespace for testing endpoint connectivity.

To reduce memory consumption, the new server in the new binary is based
on net/http instead of go-openapi and comes with a small client as well.
In total, this change reduces the resident set size (RSS) of the endpoint
process by roughly 12-15MB.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro commented on cilium-health/probeserver/main.go:
```diff
@@ -0,0 +1,84 @@
+	}
+}
+
+func writePidfile(path string) error {
```
Reviewer note: I decided to not use the `pkg/pidfile` package here, as I found that it almost doubles the RSS of the process (because it pulls in quite a few other dependencies) to ~13MB.

joestringer replied:
o_o;  Maybe `pkg/pidfile` should go on a diet? If we strip out the logging dependency, does that address this?

gandro replied:
Okay, I think I mixed up binary size and rss in my notes last week. `pidfile` isn't that bad in terms of memory, it just bloats the binary size:

The current version of the `pkg/pidfile` itself adds 1.8MB of RSS to the responder (from 5.8MB to 7.6MB in total), but adds almost 8MB to the binary size (7.3MB vs 13MB). All the overhead seems all caused by logging. Removing the logging statements brings sizes back down to normal for the responder (i.e. binary size 7.3MB, rss 5.7MB).

joestringer commented on cilium-health/launch/endpoint.go:
```diff
@@ -107,17 +108,15 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
 
-// Client wraps a client to a specific cilium-health endpoint instance, to
-// provide convenience methods such as PingEndpoint().
+// Client wraps a probe client
```
nit: The rest of the old comment is still valid, the new comment seems entirely pointless to me (the code immediately below already tells this very succinctly :-) )

gandro replied:
I fully agree! Not even sure why I edited that comment :)

joestringer commented on pkg/health/server/server.go:
```diff
@@ -264,48 +241,27 @@ func (s *Server) Shutdown() {
 	api.Logger = log.Printf
 
-	// /hello
+	// /hello - this is still used to check if admin API is up
```
Is it? 

`/hello` should be used for network-level pings. There's no need to do a ping over the unix socket, is there?

EDIT: I see you note that in the PR message, fair enough, that can be dropped in a subsequent PR.

joestringer commented on cilium-health/probeserver/Makefile:
```diff
@@ -0,0 +1,18 @@
+include ../../Makefile.defs
+
+TARGET=cilium-health-probe
```
nit: If I see a process in my process list called `cilium-health-probe`, I might think that this process is actively probing other things. But it's really just a simple HTTP server respoding to `/health`. Do you think something like `cilium-health-responder` would be more clear? (Other parts of the PR would need updating for such a change, too).

joestringer reviewed (changes requested):
> The new binary reduces the resident set size (RSS) of the endpoint process by roughly 12-15MB.

Nice! What's the effect on the container image size?

I have a few minor nits below, nothing too big. Looks good overall. I would be interested in seeing whether we could actually retain the sharing of code in `pkg/pidfile` if the logging dependency could be dropped. But that could also be addressed in a follow-up PR.

aanm commented on cilium-health/Makefile:
```diff
@@ -1,18 +1,31 @@
 	$(INSTALL) -m 0755 $(TARGET) $(DESTDIR)$(BINDIR)
+
+.PHONY: force
+force :;
```
missing new line here?

aanm reviewed (commented):
looks super great!

gandro committed:
health: Stop cilium-health instance before starting a new one

The restart functionality of the launcher was unused. This commit
ensures that the previous cilium-health instance is killed before the
launcher tries to start it again.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
